### ISSUE TITLE
Fonts gone crazy after installing v2_018 on Jetbrains [Java renderer]

### ISSUE THREAD

**Issue by rahilwazir** [chrissimpkins/Hack#152]
Fonts gone crazy after installing the `v2.018` but its working fine with `v2.010`, I did restart Jetbrain IDE. Increasing/Decreasing font have still same effect.

![crazy-fonts](https://cloud.githubusercontent.com/assets/1818941/10996531/c197ef94-84a7-11e5-93fa-1ca094fcb72b.png)

**Comment by chrissimpkins** [comment]
Thanks for reporting it.  Can you give us your platform, the editor, and indicate whether you are using the ttf or otf builds?

**Comment by rahilwazir** [comment]
Sure, here you go
- Win7 SP1 64bit
- Webstorm 10.0.4
- Using `.ttf` [provided in readme](https://github.com/chrissimpkins/Hack#recommended-binaries-for-linux-and-windows-users)

**Comment by rahilwazir** [comment]
Not sure should I open a new issue for this or not but using `v2.010` shows same wierd fonts on multiline comments. when writing `/**` (double asterik)

Example:

![code-comments-crazy-fonts](https://cloud.githubusercontent.com/assets/1818941/10997206/104f3eee-84ad-11e5-8260-2f2121ee4ec7.gif)

**Comment by chrissimpkins** [comment]
We did address numerous issues in the JetBrains renderer + Hack typeface combination after v2.010 was released.  I don’t think that this requires a separate issue report unless you are still seeing this in the current release.  Thanks for bringing it to our attention though.

**Comment by chrissimpkins** [comment]
We will take a look into the initial issue that you reported. This was a direct upgrade from the v2.010 to the v2.018 release?

**Comment by rahilwazir** [comment]
Yes It was

**Comment by chrissimpkins** [comment]
thanks!

**Comment by fruel** [comment]
Did a direct update from v2.010 to the v2.018 today too.
Visual Studio, Sublime Text 3 and all other editors are fine but all JetBrains IDEs are messed up (Android Studio 1.5, PHPStorm 10, IntelliJ 15).

System:
Windows 10 Pro x64
Installed TTF version.

Never had any rendering issues prior to v2.018.

**Comment by chrissimpkins** [comment]
> are messed up

Can you be more specific?  Are you seeing character changes like in the OP image?

**Comment by fruel** [comment]
Sorry forgot - pretty much the same as OP reported it.

Example: Laravel composer.json file
![image](https://cloud.githubusercontent.com/assets/1745966/11458312/bda3ff68-96bd-11e5-925c-d42c95176606.png)

**Comment by chrissimpkins** [comment]
Seems to be some strange Windows + JB renderer-specific issue.  Here is where I am in PyCharm 4.5 (old version) and the current PyCharm 5.0.2 release:

**PyCharm 4.5**

<img width="828" alt="pycharm4" src="https://cloud.githubusercontent.com/assets/4249591/11458419/8ecc51fc-968d-11e5-94e5-a63ea498ba7c.png">

**PyCharm 5.0.2**

<img width="688" alt="pycharm5" src="https://cloud.githubusercontent.com/assets/4249591/11458421/972dd91a-968d-11e5-8b95-5ba588c33fd1.png">


My understanding is that PyCharm uses the same Java based text renderer.  I modified the text size to 12 and have line spacing at 1.0.  This does not seem to influence the display on OS X.   

Are either of you willing to submit an issue report on the JetBrains tracker to investigate this?  I don't develop on a Windows platform and it will be difficult for me to work through the issue with them.  I'd be happy to follow the thread, pitch in information from the standpoint of our fonts (including OpenType table settings if that is where this winds up existing) and implement any changes that they recommend if they are able to identify what is going on with the renderer in the Windows editors.

**Comment by chrissimpkins** [comment]
Will take a look at a diff of our OpenType table data between the 2.010 and 2.018 versions this week as well.  Perhaps this will lead to some useful information.  

If you are willing to take the time, would it be possible to check each of the release versions between 2.010 and 2.018 to determine where we introduced the problem?  This might point us in the direction of an answer if it lies in the OpenType table settings.

**Comment by fruel** [comment]
It seems to be a general issue with the Java font renderer - not specific to JetBrains IDEs.

I wrote a quick test application using the generic Java Swing GUI system. It print the text "Hello World" twice - one the left using Hack and on the right using the default monospaced font.

![image](https://cloud.githubusercontent.com/assets/1745966/11458576/d16875f6-96c2-11e5-8ec8-e6043cc91e78.png)

Sourcecode: https://gist.github.com/fruel/1e4471b5474708f9c3e1

Maybe someone else can confirm that other Java-based application are affected too.

**Comment by fruel** [comment]
Tested the old versions with the test application from the comment above. The issue starts with v2.017. 
v2.015 and v2.016 (from here https://github.com/chrissimpkins/Hack/issues/126) are fine.

**Comment by chrissimpkins** [comment]
@fruel Thank you!  I will take a look at our OpenType table changes between those versions.  

Are you aware of other text editors in common use that use this renderer?

**Comment by chrissimpkins** [comment]
Are the editors here being released with a bundled JDK?  That seems to be the case on OS X.

**Comment by fruel** [comment]
I tested a few other applications with v2.0.17:

**jEdit**
http://sourceforge.net/projects/jedit/
Is affected by the bug and has definitely no bundled JRE.

**eclipse*
Version: Mars.1
Not affected
Don't think it has a bundeled JRE.

**NetBeans 8.1"
Affected
Not sure if JRE is bundled.

I think NetBeans and jEdit use the same GUI system whereas eclipse uses something different. (NetBeans and jEdit have similar look & feel and the font-selection dialogs are identical)
 
Wikipedia confirms this:
```
Eclipse implements the graphical control elements of the Java toolkit called SWT, whereas most Java applications use the Java standard Abstract Window Toolkit (AWT) or Swing. 
```

**Comment by chrissimpkins** [comment]
@fruel Thank you very much for all of this information.  This is extremely helpful.  I will post some information about the differences in our OpenType tables between the identified versions this week.

**Comment by fruel** [comment]
New discovery:

I have multiple Java versions installed and only older ones are affected. When installing NetBeans it seems the installer changed the default Java runtime as Hack v2.018 is now working everywhere except in NetBeans.

I tested this manually with the my test application from above:
v2.018 works fine in my test application with JRE 1.8u66 x64 and JDK 1.8u66 x64 but not with JDK 1.8u51 x64 or JDK 1.7u55 x86

What Java version do you have on OSX?

**Comment by chrissimpkins** [comment]
```
java version "1.8.0_05"
Java(TM) SE Runtime Environment (build 1.8.0_05-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.5-b02, mixed mode)
```

**Comment by fruel** [comment]
OK...this thing is getting really weird....

If I start my test application from the command line using 1.8u51 it does not work but IntelliJ claims to be running with 1.8u51 and it is working now:

![image](https://cloud.githubusercontent.com/assets/1745966/11458975/39aff84c-96cc-11e5-98de-45a06039b806.png)

![image](https://cloud.githubusercontent.com/assets/1745966/11458982/7be673c6-96cc-11e5-99b4-352466cb71e2.png)

**Comment by chrissimpkins** [comment]
PyCharm v5.0.2 on OS X: 

<img width="377" alt="jre-version" src="https://cloud.githubusercontent.com/assets/4249591/11459099/c6bb5e64-969c-11e5-9d3b-53be2f837568.png">

Different than the installed version on my PATH.

**Comment by chrissimpkins** [comment]
Please reopen if this remains an issue for you.

**Comment by rahilwazir** [comment]
I'm not sure why did you close this issue, but this issue still exist with latest Webstorm `v11.0.2`

**Comment by chrissimpkins** [comment]
@rahilwazir More than happy to continue to help here though I am not on Windows and cannot reproduce the issue on OS X versions of JetBrains editors.  

@fruel would you be willing to release that code that you used to render fonts with Java?  This would be an extremely helpful tool for us and might help us locate the problematic JRE versions.  Perhaps with this information, one of the two of you could submit an issue report on the JetBrains issue tracker so that we can explore this in more detail.

**Comment by fruel** [comment]
You can find the source code for my demo application linked im my comment from Nov. 29. It's a basic java window.

It still does not work for me in Android Studio but it works in PHPStorm - I'm not sure why or what version of java are affected as it behaved really weird.

**Comment by chrissimpkins** [comment]
@fruel thanks! sorry, I missed the link.  

@fruel & @rahilwazir:  Here is where I am on OS X with Java(TM) SE Runtime Environment (build 1.8.0_05-b13) vs. Source Code Pro:

<img width="485" alt="java1" src="https://cloud.githubusercontent.com/assets/4249591/11845631/c7367f9c-a3e2-11e5-9236-5bc52b225f55.png">

and with Java(TM) SE Runtime Environment (build 1.8.0_66-b17):

<img width="478" alt="java2" src="https://cloud.githubusercontent.com/assets/4249591/11845677/05a8ff5c-a3e3-11e5-8bc7-ca038139101f.png">

**Comment by chrissimpkins** [comment]
@rahilwazir can you link the JRE that your version of WebStorm is using on Windows?

**Comment by chrissimpkins** [comment]
@fruel possible to modify this script to read in a text specimen file and render on a scrollable display window with Java?  I'd like to turn this into a tool for typeface developers.  Interested?

**Comment by fruel** [comment]
"I'm guessing that you simply switch to Windows paths for the compiled binary root directory with the -cp switch on Windows" - Yes, same commands - only with Windows paths.

Demo application:
I can't promise anything as I've many other things to do right now but what would you need?
Loading a textfile and a TTF file and a window to display the text in the given font?

**Comment by chrissimpkins** [comment]
Thank you.  Yes that description fits.  I would need some help with the file I/O and Swing UI.  Not a Java developer and would like to have something simple that will test rendering with Java since these JetBrains editors have caused us a significant number of problems and are so widely used.  This was a great idea that you had and I think it will help with tests out of the editor with other Java versions.  Do you happen to know if the Font class supports a setting to apply and not apply True Type instructions (hinting) in the rendered fonts?

**Comment by chrissimpkins** [comment]
and my scrollable comment simply meant that we will need to be able to support reasonably large blocks of specimen text so the UI window should support text overflow with a scrollbar.

**Comment by fruel** [comment]
Ok - I might be able write it on Saturday or Sunday. 

About the hinting: Not sure. Have to check the docs.

**Comment by chrissimpkins** [comment]
Absolutely no rush at all.  I will take a look through the documentation too.

**Comment by chrissimpkins** [comment]
@rahilwazir Rahil, will you be able to provide the information above about the success with the Java versions on your system?  This should help us to better understand this problem.

**Comment by chrissimpkins** [comment]
@fruel let’s move the conversation about the tool to #161.  We will continue to discuss @rahilwazir's original issue here.

**Comment by rahilwazir** [comment]
@chrissimpkins Sorry for late reply, here you go:

![screenshot from 2015-12-18 12 44 42](https://cloud.githubusercontent.com/assets/1818941/11891923/34970324-a585-11e5-901d-0b6db2b28a92.png)

Compiled with 
```
$ javac -version
Picked up _JAVA_OPTIONS: -Dawt.useSystemAAFontSettings=lcd -Dsun.java2d.xrender=true
javac 1.8.0_45
```

**Comment by rahilwazir** [comment]
Sorry for the change in status, I had posted the result of Linux machine mistakenly

Anyway here is the result (Compiled with `javac 1.8.0_65`):

![hack_java_font](https://cloud.githubusercontent.com/assets/1818941/11892466/266e9b0e-a58a-11e5-9d1c-475308819256.png)

**Comment by chrissimpkins** [comment]
@rahilwazir it looks like it is time to go to JetBrains for more information.  That is the same version as the reported version in WebStorm and it renders fine.  Mind submitting your issue in their issue tracker and update us with the outcome?  Perhaps they modified the Java font rendering code in some way?

**Comment by rahilwazir** [comment]
I have posted the issue, just so you know:
https://youtrack.jetbrains.com/issue/IDEA-149554

I'll update here if I get any reply from them

**Comment by yoyosan** [comment]
Hi @rahilwazir !

I've had this issue too and it was because I've installed a newer version of the Hack font without removing the existing one.

Also, you need to restart your IDE.

**Comment by fruel** [comment]
While working on the test tool for #161 i discovered that the fonts on my PC are messed up. It looks to me that there were multiple versions of Hack installed as some applications were using C:\Windows\Fonts\Hack_Regular.ttf and some C:\Windows\Fonts\Hack_Regular**_0**.ttf

I removed Hack from the Windows fonts folder (had to do it multiple times as the file reappeared) - but some Java applications are still able to use it. 

TL:DR: 
Windows is messed up. Might explain the confusing results in my previous comments. Will try it on another computer.

**Comment by chrissimpkins** [comment]
@rahilwazir Thanks Rahil! I will keep an eye on it as well.

**Comment by chrissimpkins** [comment]
@yoyosan Thanks for this information Marius.  Really appreciate it.

**Comment by chrissimpkins** [comment]
> Windows is messed up. Might explain the confusing results in my previous comments. Will try it on another computer.

Interesting.  I wonder why it would lead to a render with incorrect glyphs rather than simply selecting one or the other set of font files?  Perhaps this belongs on a Windows issue tracker?

**Comment by chrissimpkins** [comment]
@fruel Thanks for creating the Java font testing tool by the way.  It looks great!

**Comment by rahilwazir** [comment]
@yoyosan Thanks that actually worked, but this is so weird, not most of the users remove fonts when they upgrade.

@chrissimpkins Can you please make a note of this update process on readme for future readers. I'll close this once you do.

**Comment by chrissimpkins** [comment]
I'd be glad to.  What approach did you use to delete them prior to the install?

**Comment by rahilwazir** [comment]
Just navigate to `c:\windows\fonts`, search for the `Hack` font and delete, then install the newer version.

**Comment by yoyosan** [comment]
@chrissimpkins Go to Control Panel\Appearance and Personalization\Fonts(Windows 7), search for Hack, right click and Delete.

**Comment by chrissimpkins** [comment]
@rahilwazir @yoyosan both of these approaches fix the issue?

**Comment by chrissimpkins** [comment]
And is there a reboot involved?  I saw that the editor needs to be closed and reopened in the comments below.

**Comment by rahilwazir** [comment]
I'm not sure about the [yoyosan](https://github.com/chrissimpkins/Hack/issues/152#issuecomment-166828747) solution (as I'm not currently on my Windows machine).
Yes the editor need to be restarted.

**Comment by chrissimpkins** [comment]
thank you @rahilwazir !

**Comment by rahilwazir** [comment]
@chrissimpkins Yeah both approaches works, @yoyosan solution is other way around

**Comment by chrissimpkins** [comment]
I will add this to the README and let you know when it is available. Thanks again. Odd problem.

**Comment by chrissimpkins** [comment]
README updated with workaround for this issue.  Users are still having the problem and the issue report remains open on the JetBrains tracker.  Will continue to leave the issue open here for reference.

**Comment by rahilwazir** [comment]
@chrissimpkins The [issue](https://youtrack.jetbrains.com/issue/IDEA-149554) is closed now

**Comment by chrissimpkins** [comment]
Alright, I will close this issue.  Thanks for raising it @rahilwazir

**Comment by texhex** [comment]
@fruel I know this issue is closed, but just for the sake of documentation: The issue that you suddenly had a C:\Windows\Fonts\Hack_Regular.ttf and some C:\Windows\Fonts\Hack_Regular**_0**.ttf file is caused by the ***Fonts*** applet. 

As soon as you open C:\Windows\Fonts (or use ***Fonts*** from Control Panel) and then drag&drop a TTF file that is already there and/or locked, Windows will decide to copy the file as OrignalFileName**_0**.ttf. This doesn't make any sense, but that's what the coder of the Microsoft.Fonts applet did.

**Comment by chrissimpkins** [comment]
@texhex Thank you for this information Michael.  Appreciate the feedback.  Odd that the duplicates that include different file paths lead to such problems...

**Comment by rahilwazir** [comment]
@texhex Nice to know, thanks for the inside

**Comment by ragingnerd** [comment]
Does anyone have this this issue after upgrading to webstorm 2016?

**Comment by rahilwazir** [comment]
@ragingnerd Well you are correct, I forgot to comment here.

@chrissimpkins Can you look into this? PHPStorm 2016 causes this behavior again, even deleting old fonts before installing the new one.

OS: Windows 10 x64, IDE: PHPStorm 2016.1

**Comment by chrissimpkins** [comment]
@ragingnerd @rahilwazir 

Can you both confirm that you do not have duplicate fonts as a result of the Fonts applet that @texhex reported in https://github.com/chrissimpkins/Hack/issues/152#issuecomment-181074350?

**Comment by ragingnerd** [comment]
@chrissimpkins 

I think that might be the root cause. Font properties for the individual types show up as:

- Hack-BoldItalic_0.ttf
- Hack-Italic_1.ttf
- Hack-Bold_2.ttf
- Hack-Regular_3.ttf

but all my system fonts have a suffix of _3 (Eg. arial_3.ttf)

I've tried uninstalling the font, doing a clean reboot and reinstalling the fonts by clicking on them individually, then invalidating cache and restarting the ide.


**OS**: Windows 8.1 x64
**IDE**: WebStorm 2016.1

**Comment by chrissimpkins** [comment]
@ragingnerd any idea how those numbers are being applied?  Is that install order or is Windows supplementing the file path with variant meta info through the number?   This numbering system in the file paths (on clean installs) seems to be a new Windows issue.  I wonder if we are dealing with a change in Windows rather than the JetBrains editors.

**Comment by ragingnerd** [comment]
@chrissimpkins no idea at this stage. I've tried reinstalling 2.019 and 2.018 .TTF fonts, the number of times / order I install the fonts in have no bearing on the numbers. I've also tried installing the Hack OTF fonts - no suffixes.

**Comment by chrissimpkins** [comment]
@ragingnerd thank you for checking!  do you see this problem with the otf builds?

**Comment by ragingnerd** [comment]
@chrissimpkins no problem! 2.019 .OTF Bold is fine, the other 3 have _0 appended to them - IDE renders everything in bold.

**Comment by chrissimpkins** [comment]
@texhex I think we may need your expertise here.  Any thoughts about these issues with the Windows installs?

**Comment by texhex** [comment]
@chrissimpkins OK, I will do my very best. Let's see if this is enough :)

@ragingnerd First, we should remove anything "Hack" related in order to have a clean state. Please first go to the Fonts applet (Right-click on Start menu -> Control Panel -> Appearance -> Fonts). Search for Hack and hit delete. Windows should then delete the fonts. If it says something about **File XXX in use, can't delete**  don't worry. Next, restart the machine. Trust me, Windows because the caching requires this. Once the machine is back, fire up `Regedit.exe` and move to 

**HKEY_LOCAL_MACHINE` -> SOFTWARE -> Microsoft -> Windows NT -> CurrentVersion -> Fonts**

Inside this, locate the entries ("Name") starting with Hack (TrueType) etc. and delete them. Also, check if the column "Data" lists anywhere any Hack_XX.ttf entries. If so, delete them.

Finally, below "Fonts" inside `regedit.exe` (left side) is another entry called **FontSubstitutes**. Check if there are any entries listed with "Hack". If so, delete them.

Restart the machine.

Move back to the Fonts applet and check if there is still something "Hack" related listed. 
If so delete it and recheck the registry again. There shouldn't be anything, but you never now.

Finally, we do a check if really a files are gone. Open CMD.EXE and execute `dir "C:\Windows\Fonts\Hack*.*"`. There should be no result. If there are files listed, delete them. 

Restart the machine one final time. Recheck WebStorm. It shouldn't list Hack anymore and using any other font should be OK. If it still lists Hack, please let me know. In this case we might have the rare case of a damaged Font Cache 3.0 which requires stopping the "Windows FontCache" service and deleting the "FontCache3.0*.dat" file in **C:\Windows\ServiceProfiles\LocalService\AppData\Local**.  

Then try to install Hack again by drag&drop the TTF file to the fonts applet. And you know it already: Restart the machine after that before trying to use the font.

That's all I can think of right now. HTH!

P.S.: If you think the font handling of Windows is b0rken by design: I couldn't agree more :).

**Comment by chrissimpkins** [comment]
@texhex Wow, that is a convoluted mess on Windows.  It sounds like Windows doesn’t intend to have fonts updated very often.  Thanks so much for all of this information.  @rahilwazir @ragingnerd please let us know if the above works for you.

**Comment by texhex** [comment]
@chrissimpkins Yes, it's a total mess. The main problem is that Windows has several "data stores" (Files, Registry, Cache) and no sanity check in place if this all makes sense. I would assume that single coder, that has the full backup from the Windows lead, could get all this sorted out in less than two weeks or so. But hey, it only took them two Windows version to bring the start menu back :-).

**Comment by rahilwazir** [comment]
@chrissimpkins Well that actually works :smile: 

Thanks @texhex once again, so weird shit going on

**Comment by chrissimpkins** [comment]
@texhex :+1: !!

**Comment by ragingnerd** [comment]
@texhex  @chrissimpkins All sorted, thanks for your help!

Btw for anyone having the same problem, when I do a  `dir "C:\Windows\Fonts\Hack*.*"`, there are hack files but they are hidden on the font applet. I had to `del dir "C:\Windows\Fonts\Hack*.*"` on an elevated cmd to get rid of it.

**Comment by texhex** [comment]
@ragingnerd Good to hear. Finally you can enjoy the most awesome font ever :-). 

Blunt self promotion: I have create a setup.exe for Hack (and some other fonts) that should make installation on a new machine rather easy. It takes also care about those Hack_0/1/2.ttf files and about the Font Cache service(s). You might want to give it a try for the next machine: [OS Font Pack](https://github.com/texhex/OSFontPack). 

Note: This setup is not approved by @chrissimpkins, but I hope some day it will.

**Comment by chrissimpkins** [comment]
> I hope some day it will

I will try to get the docs updated this week.  Sorry for the delay and thanks again!

**Comment by texhex** [comment]
@chrissimpkins No rush needed Chris, I just wanted to make sure that the setup is not "official". The message wasn't intend to push you to anything. Really.

**Comment by alicerinky-lab** [comment]
danishmir729 _id hack insetragrame
